core/Enclave/enclave_hsm.cpp

```diff
@@ -147,653 +157,2072 @@ static uint32_t sgx_get_gcm_ciphertext_size(const sgx_aes_gcm_data_ex_t *gcm_dat
+        default:
+            return UINT32_MAX;
     }
-
-    return ret;
 }
 
-static sgx_status_t sgx_gcm_decrypt(const sgx_aes_gcm_128bit_key_t *key,
-                                    uint32_t *plaintext_size, uint8_t *plaintext,
-                                    const sgx_aes_gcm_data_ex_t *gcm_data)
+static uint32_t ehsm_get_rsa_public_key_pem_size(const uint32_t keyspec)
 {
-    if (NULL == gcm_data || NULL == plaintext || NULL == *plaintext_size
-                 || *plaintext_size < sgx_get_gcm_ciphertext_size(gcm_data))
-        return SGX_ERROR_INVALID_PARAMETER;
-
-    sgx_status_t ret = sgx_rijndael128GCM_decrypt(key,
-                                                  gcm_data->payload, gcm_data->ciphertext_size,
-                                                  plaintext,
-                                                  gcm_data->iv, sizeof(gcm_data->iv),
-                                                  &(gcm_data->payload[gcm_data->ciphertext_size]), gcm_data->aad_size,
-                                                  (const sgx_aes_gcm_128bit_tag_t*)gcm_data->mac);
-    if (SGX_SUCCESS != ret)
-        printf("gcm decrypting failed.\n");
-    else
-        *plaintext_size = sgx_get_gcm_ciphertext_size(gcm_data);
-
-    return ret;
+    switch(keyspec)
+    {
+        case EH_RSA_2048:
+            return RSA_2048_PUBLIC_KEY_PEM_SIZE;
+        case EH_RSA_3072:
+            return RSA_3072_PUBLIC_KEY_PEM_SIZE;
+        case EH_RSA_4096:
+            return RSA_4096_PUBLIC_KEY_PEM_SIZE;
+        default:
+            return 0;
+    }
 }
 
-
-sgx_status_t enclave_create_aes_key(uint8_t *cmk_blob, uint32_t cmk_blob_size, uint32_t *req_blob_size)
+// sgx-ssl-framework start
+
+/* Rijndael AES-GCM
+* Parameters:
+*   Return: sgx_status_t  - SGX_SUCCESS or failure as defined sgx_error.h
+*   Inputs: sgx_aes_gcm_128bit_key_t *p_key - Pointer to key used in encryption/decryption operation
+*           uint8_t *p_src - Pointer to input stream to be encrypted/decrypted
+*           uint32_t src_len - Length of input stream to be encrypted/decrypted
+*           uint8_t *p_iv - Pointer to initialization vector to use
+*           uint32_t iv_len - Length of initialization vector
+*           uint8_t *p_aad - Pointer to input stream of additional authentication data
+*           uint32_t aad_len - Length of additional authentication data stream
+*           sgx_aes_gcm_128bit_tag_t *p_in_mac - Pointer to expected MAC in decryption process
+*   Output: uint8_t *p_dst - Pointer to cipher text. Size of buffer should be >= src_len.
+*           sgx_aes_gcm_128bit_tag_t *p_out_mac - Pointer to MAC generated from encryption process
+* NOTE: Wrapper is responsible for confirming decryption tag matches encryption tag */
+sgx_status_t aes_gcm_encrypt(const uint8_t *p_key, const uint8_t *p_src, uint32_t src_len,
+                             uint8_t *p_dst, const uint8_t *p_iv, uint32_t iv_len, const uint8_t *p_aad,
```

Comment:
seems this function not necessary, and many duplicated checks with the caller.

and also this function just called by the enclave_aes_encrypt(), so you can put the logic directly into enclave_aes_encrypt()

Reply:
OK, I'll put them into enclave_aes_encrypt() directly.